include/seastar/core/memory.hh

```diff
@@ -382,29 +383,76 @@ public:
     void operator=(scoped_large_allocation_warning_disable&&) = delete;
 };
 
-/// Enable/disable heap profiling.
+/// Describes an allocation location in the code.  The location is identified by
+/// its backtrace. One allocation_site can represent many allocations at the
+/// same location. `count` and `size` represent the cumulative sum of all
+/// allocations at the location.
+struct allocation_site {
+    mutable size_t count = 0; // number of live objects allocated at backtrace.
+    mutable size_t size = 0; // amount of bytes in live objects allocated at backtrace.
+    simple_backtrace backtrace; // call site for this allocation
+
+    // All allocation sites are linked to each other. This can be used for easy
+    // iteration across them in gdb scripts where it's difficult to work with
+    // the C++ data structures.
+    mutable const allocation_site* next = nullptr; // next allocation site in the chain
+    mutable const allocation_site* prev = nullptr; // previous allocation site in the chain
+
+    bool operator==(const allocation_site& o) const {
+        return backtrace == o.backtrace;
+    }
+
+    bool operator!=(const allocation_site& o) const {
+        return !(*this == o);
+    }
+};
+
+/// If memory sampling is on returns the current sampled memory live set
+std::vector<allocation_site> sampled_memory_profile();
+
+/// Copies the current sampled set of allocation_sites in the output parameter
+/// (up to size of output vector) Returns amount of copied elements. Useful if
+/// one wants to avoid allocating an output vector (e.g.: under OOM conditions)
+size_t sampled_memory_profile(std::vector<allocation_site>&);
+
+/// Enable sampled heap profiling by setting a sample rate
+/// disable heap profiling by setting the sample rate to 0
 ///
 /// In order to use heap profiling you have to define
 /// `SEASTAR_HEAPPROF`.
-/// Heap profiling data is not currently exposed via an API for
-/// inspection, instead it was designed to be inspected from a
-/// debugger.
+///
+/// Use \ref sampled_memory_profile for API access to profiling data
+///
 /// For an example script that makes use of the heap profiling data
 /// see [scylla-gdb.py] (https://github.com/scylladb/scylla/blob/e1b22b6a4c56b4f1d0adf65d1a11db4bcb51fe7d/scylla-gdb.py#L1439)
 /// This script can generate either textual representation of the data,
 /// or a zoomable flame graph ([flame graph generation instructions](https://github.com/scylladb/scylla/wiki/Seastar-heap-profiler),
 /// [example flame graph](https://user-images.githubusercontent.com/1389273/72920437-f0cf8a80-3d51-11ea-92f0-f3dbeb698871.png)).
-void set_heap_profiling_enabled(bool);
```

Comment:
So this "replaces" the old sampling mode of sample-everything to be inspected in gdb, right?

This might not fly upstream, we'll have to see. Maybe let's focus on the RP side of things for now.

Reply:
> So this "replaces" the old sampling mode of sample-everything to be inspected in gdb, right?

The gdb still works

 > This might not fly upstream, we'll have to see. Maybe let's focus on the RP side of things for now.

I very much agree with this. IMO I'd prefer to just remove the GDB way (as it means we could get rid of those extra allocation site links) but at the end of the day upstream will decide and I don't think there is much point in guess in now.